File: sdk-go/processes_test.go
package boxkite

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"testing"
)

func TestStartProcess_DefaultsMaxRuntimeSecondsTo3600(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sandboxes/sess-1/processes" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["max_runtime_seconds"] != float64(3600) {
			t.Errorf("expected default max_runtime_seconds=3600, got %v", body["max_runtime_seconds"])
		}
		if body["command"] != "npm run dev" {
			t.Errorf("unexpected command: %v", body["command"])
		}
		writeJSON(t, w, 201, `{"process_id": "proc-1", "status": "running", "started_at": "2026-01-01T00:00:00Z"}`)
	})
	defer closeServer()

	result, err := client.StartProcess(context.Background(), "sess-1", "npm run dev", nil)
	if err != nil {
		t.Fatalf("StartProcess: %v", err)
	}
	if result.ProcessID != "proc-1" || result.Status != "running" {
		t.Errorf("unexpected result: %+v", result)
	}
}

func TestStartProcess_CustomOptions(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["max_runtime_seconds"] != float64(1800) {
			t.Errorf("unexpected max_runtime_seconds: %v", body["max_runtime_seconds"])
		}
		if body["description"] != "dev server" {
			t.Errorf("unexpected description: %v", body["description"])
		}
		if body["expose_port"] != float64(3000) {
			t.Errorf("unexpected expose_port: %v", body["expose_port"])
		}
		writeJSON(t, w, 201, `{"process_id": "proc-2", "status": "running", "started_at": "2026-01-01T00:00:00Z"}`)
	})
	defer closeServer()

	_, err := client.StartProcess(context.Background(), "sess-1", "npm run dev", &StartProcessOptions{
		Description:       Ptr("dev server"),
		MaxRuntimeSeconds: 1800,
		ExposePort:        Ptr(3000),
	})
	if err != nil {
		t.Fatalf("StartProcess: %v", err)
	}
}

func TestListProcesses(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/v1/sandboxes/sess-1/processes" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		writeJSON(t, w, 200, `{"processes": [{"process_id": "proc-1", "command": "npm run dev", "description": null, "status": "running", "started_at": "2026-01-01T00:00:00Z", "exit_code": null}]}`)
	})
	defer closeServer()

	result, err := client.ListProcesses(context.Background(), "sess-1")
	if err != nil {
		t.Fatalf("ListProcesses: %v", err)
	}
	if len(result.Processes) != 1 || result.Processes[0].ProcessID != "proc-1" {
		t.Errorf("unexpected result: %+v", result)
	}
}

func TestGetProcessOutput_SendsSinceOffset(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.URL.Query().Get("since_offset"); got != "27" {
			t.Errorf("expected since_offset=27, got %q", got)
		}
		writeJSON(t, w, 200, `{"status": "running", "stdout_chunk": "more\n", "next_offset": 32, "truncated": false, "exit_code": null}`)
	})
	defer closeServer()

	result, err := client.GetProcessOutput(context.Background(), "sess-1", "proc-1", 27)
	if err != nil {
		t.Fatalf("GetProcessOutput: %v", err)
	}
	if result.NextOffset != 32 {
		t.Errorf("unexpected next_offset: %d", result.NextOffset)
	}
}

func TestSendProcessInput(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sandboxes/sess-1/processes/proc-1/input" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["data"] != "y\n" {
			t.Errorf("unexpected data: %v", body["data"])
		}
		writeJSON(t, w, 200, `{"bytes_written": 2}`)
	})
	defer closeServer()

	result, err := client.SendProcessInput(context.Background(), "sess-1", "proc-1", "y\n")
	if err != nil {
		t.Fatalf("SendProcessInput: %v", err)
	}
	if result.BytesWritten != 2 {
		t.Errorf("unexpected bytes_written: %d", result.BytesWritten)
	}
}

func TestStopProcess(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sandboxes/sess-1/processes/proc-1/stop" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"status": "stopped", "exit_code": 143}`)
	})
	defer closeServer()

	result, err := client.StopProcess(context.Background(), "sess-1", "proc-1")
	if err != nil {
		t.Fatalf("StopProcess: %v", err)
	}
	if result.Status != "stopped" || result.ExitCode == nil || *result.ExitCode != 143 {
		t.Errorf("unexpected result: %+v", result)
	}
}

File: sdk-rust/tests/processes_test.rs
mod common;

use boxkite_client::StartProcessOptions;
use serde_json::json;
use wiremock::matchers::{body_json, method, path, query_param};
use wiremock::{Mock, ResponseTemplate};

#[tokio::test]
async fn start_process_sends_only_set_optional_fields() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess-1/processes"))
        .and(body_json(json!({
            "command": "npm run dev",
            "max_runtime_seconds": 3600
        })))
        .respond_with(ResponseTemplate::new(201).set_body_json(json!({
            "process_id": "proc_1", "status": "running", "started_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    let result = client
        .start_process("sess-1", "npm run dev", StartProcessOptions::new())
        .await
        .expect("start_process should succeed");

    assert_eq!(result.process_id, "proc_1");
    assert_eq!(result.status, "running");
}

#[tokio::test]
async fn start_process_sends_all_set_optional_fields() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess-1/processes"))
        .and(body_json(json!({
            "command": "npm run dev",
            "description": "dev server",
            "max_runtime_seconds": 1800,
            "expose_port": 3000
        })))
        .respond_with(ResponseTemplate::new(201).set_body_json(json!({
            "process_id": "proc_2", "status": "running", "started_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    let options = StartProcessOptions::new()
        .description("dev server")
        .max_runtime_seconds(1800)
        .expose_port(3000);

    client
        .start_process("sess-1", "npm run dev", options)
        .await
        .expect("start_process should succeed");
}

#[tokio::test]
async fn list_processes_returns_all_tracked_processes() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/sandboxes/sess-1/processes"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "processes": [
                {
                    "process_id": "proc_1", "command": "npm run dev", "description": null,
                    "status": "running", "started_at": "2026-01-01T00:00:00Z",
                    "exit_code": null, "expose_port": 3000
                },
                {
                    "process_id": "proc_2", "command": "sleep 1", "description": "quick nap",
                    "status": "exited", "started_at": "2026-01-01T00:00:00Z",
                    "exit_code": 0, "expose_port": null
                }
            ]
        })))
        .mount(&server)
        .await;

    let result = client
        .list_processes("sess-1")
        .await
        .expect("list_processes should succeed");

    assert_eq!(result.processes.len(), 2);
    assert_eq!(result.processes[0].expose_port, Some(3000));
    assert_eq!(result.processes[1].expose_port, None);
    assert_eq!(result.processes[1].exit_code, Some(0));
}

#[tokio::test]
async fn get_process_output_sends_since_offset_as_query_param() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/sandboxes/sess-1/processes/proc_1/output"))
        .and(query_param("since_offset", "42"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "status": "running", "stdout_chunk": "hello\n", "next_offset": 48,
            "truncated": false, "exit_code": null
        })))
        .mount(&server)
        .await;

    let result = client
        .get_process_output("sess-1", "proc_1", 42)
        .await
        .expect("get_process_output should succeed");

    assert_eq!(result.stdout_chunk, "hello\n");
    assert_eq!(result.next_offset, 48);
}

#[tokio::test]
async fn send_process_input_returns_bytes_written() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess-1/processes/proc_1/input"))
        .and(body_json(json!({"data": "y\n"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({"bytes_written": 2})))
        .mount(&server)
        .await;

    let result = client
        .send_process_input("sess-1", "proc_1", "y\n")
        .await
        .expect("send_process_input should succeed");

    assert_eq!(result.bytes_written, 2);
}

#[tokio::test]
async fn stop_process_returns_exit_code() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess-1/processes/proc_1/stop"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "status": "stopped", "exit_code": 0
        })))
        .mount(&server)
        .await;

    let result = client
        .stop_process("sess-1", "proc_1")
        .await
        .expect("stop_process should succeed");

    assert_eq!(result.status, "stopped");
    assert_eq!(result.exit_code, Some(0));
}

#[tokio::test]
async fn get_process_output_maps_unknown_process_id_to_404() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/sandboxes/sess-1/processes/proc_missing/output"))
        .respond_with(ResponseTemplate::new(404).set_body_json(json!({
            "error": {"code": "not_found", "message": "Process not found"}
        })))
        .mount(&server)
        .await;

    let err = client
        .get_process_output("sess-1", "proc_missing", 0)
        .await
        .unwrap_err();

    assert_eq!(err.code(), Some("not_found"));
    assert_eq!(err.status(), Some(404));
}
